synapse/rest/admin/event_reports.py

```diff
@@ -0,0 +1,65 @@
+# -*- coding: utf-8 -*-
+# Copyright 2020 Dirk Klimpel
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+
+from synapse.http.servlet import RestServlet, parse_integer, parse_string
+from synapse.rest.admin._base import admin_patterns, assert_requester_is_admin
+
+logger = logging.getLogger(__name__)
+
+
+class EventReportsRestServlet(RestServlet):
+    """
+    List all reported events that are known to the homeserver. Results are returned
+    in a dictionary containing report information. Supports pagination.
+    This needs user to have administrator access in Synapse.
+
+    GET /_synapse/admin/v1/event_reports?from=0&limit=10
+    returns:
+        200 OK with list of reports if success otherwise an error.
+
+    Args
+        The parameters `from` and `limit` are required only for pagination.
+        By default, a `limit` of 100 is used.
+        The parameter `user_id` can be used to filter by user id.
+        The parameter `room_id` can be used to filter by room id.
+    Returns:
+        A list of reported events and an integer representing the total number of
+        reported events that exist given this query
+    """
+
+    PATTERNS = admin_patterns("/event_reports$")
+
+    def __init__(self, hs):
+        self.hs = hs
+        self.auth = hs.get_auth()
+        self.store = hs.get_datastore()
+
+    async def on_GET(self, request):
+        await assert_requester_is_admin(self.auth, request)
+
+        start = parse_integer(request, "from", default=0)
+        limit = parse_integer(request, "limit", default=100)
+        user_id = parse_string(request, "user_id", default=None)
+        room_id = parse_string(request, "room_id", default=None)
+
+        event_reports, total = await self.store.get_event_reports_paginate(
+            start, limit, user_id, room_id
+        )
+        ret = {"event_reports": event_reports, "total": total}
+        if len(event_reports) >= limit:
```

Comment:
`len(event_reports)` should hopefully never be greater than `limit`. Perhaps you meant `if start + len(event_reports) < total:`?

Reply:
You can find the same code here: https://github.com/matrix-org/synapse/blob/97962ad17b204be0a88ef0cd3026f11c359fdb4a/synapse/rest/admin/users.py#L100-L102
If ``len(event_reports)`` equals ``limit`` you will get more results on next page.

Reply:
I see, this was a bit confusing. `==` is really the case we care about here, but we're doing `>=` for safety.

So... does this mean if there are 100 reports, and we set a limit of 100, we'll get a `next_token` that doesn't actually show any more reports?

Reply:
Oh. There is a bug. You are right. I will add a test.

Reply:
I'm not sure. Should be ``next_token`` an integer or string?
In the user api it is a string:
https://github.com/matrix-org/synapse/blob/babc0275431c68e64050db11959d74a636afbd3e/synapse/rest/admin/users.py#L102

In room api an integer:
https://github.com/matrix-org/synapse/blob/babc0275431c68e64050db11959d74a636afbd3e/synapse/rest/admin/rooms.py#L205

Reply:
Generally we prefer pagination tokens to be strings, so that we can encode extra data into them if necessary.

However, considering we're not very likely to do that here, and it would be unfortunate to have one admin API with a string pagination token, and another with an int one: I'd keep it as int for now.

In the future if we need to encode extra data for some reason, we'll just bump the endpoint version.